### PR TITLE
rgw/sts: adding code for federated user as owner in case of STS.

### PR DESCRIPTION
**Pull request by pritha-srivastava**
A new user under the namespace 'oidc' is created for every federated
user in case of AssumeRoleWithWebIdentity.

In case of AssumeRole, the user that needs cross account access becomes
the owner.

In both cases buckets are created in the tenant that the role belongs to.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by pritha-srivastava** [comment]
This PR is to get some early feedback for user as an owner of resources in case of STS. The external user has been created in 'oidc' namespace. For example if a role is created in the global namespace, then the user appears as - $oidc$4ecc89ca-783e-481f-9bda-aecae1df58d7. user id 4ecc89ca-783e-481f-9bda-aecae1df58d7 is based on the 'sub' claim of the incoming access token, which is also required by the AssumeRoleWithWebIdentity call, since that is one of the output parameters.

**Comment by pritha-srivastava** [comment]
radosgw-admin user stats, bucket stats, bucket chown, usgae show have been tested and are working fine. The o/p of bucket stats is follows:
{
        "bucket": "my-bucket-1",
        "num_shards": 11,
        "tenant": "",
        "zonegroup": "9d9ad35c-727c-4d91-be6f-51832132abbd",
        "placement_rule": "default-placement",
        "explicit_placement": {
            "data_pool": "",
            "data_extra_pool": "",
            "index_pool": ""
        },
        "id": "d6c588b7-ea84-4524-98b9-e3a820cade6e.4137.2",
        "marker": "d6c588b7-ea84-4524-98b9-e3a820cade6e.4137.2",
        "index_type": "Normal",
        "owner": "4ecc89ca-783e-481f-9bda-aecae1df58d7",
        "ver": "0#1,1#1,2#2,3#1,4#2,5#1,6#1,7#1,8#1,9#1,10#1",
        "master_ver": "0#0,1#0,2#0,3#0,4#0,5#0,6#0,7#0,8#0,9#0,10#0",
        "mtime": "2020-07-23T05:34:41.911718Z",
        "creation_time": "2020-07-23T05:34:41.540354Z",
        "max_marker": "0#,1#,2#,3#,4#,5#,6#,7#,8#,9#,10#",
        "usage": {
            "rgw.main": {
                "size": 42,
                "size_actual": 8192,
                "size_utilized": 42,
                "size_kb": 1,
                "size_kb_actual": 8,
                "size_kb_utilized": 1,
                "num_objects": 2
            }
        },
        "bucket_quota": {
            "enabled": false,
            "check_on_raw": false,
            "max_size": -1,
            "max_size_kb": 0,
            "max_objects": -1
        }
    }

**Comment by mattbenjamin** [comment]
@pritha-srivastava +1 on adding namespace to output

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -3057,6 +3058,7 @@ int main(int argc, const char **argv)
 
   rgw_user user_id;
   string tenant;
+  string ns;
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc:3061]: suggest user_ns to distinguish from other namespacing (multi and other)

File: src/rgw/rgw_basic_types.h [hunk 2]
```diff
@@ -14,36 +14,51 @@ class cls_user_bucket;
...
   }
   void decode(ceph::buffer::list::const_iterator& bl) {
-    DECODE_START(1, bl);
+    DECODE_START(2, bl);
     decode(tenant, bl);
     decode(id, bl);
+    if (struct_v >= 2) {
+      decode(ns, bl);
+    } else {
+      ns = "";
```
Comment (mattbenjamin) [inline src/rgw/rgw_basic_types.h:48]: doesn't this initialize to "" without assignment?

File: src/rgw/rgw_admin.cc (hunk 1 above)
Comment (pritha-srivastava) [reply src/rgw/rgw_admin.cc:3061]: ok

**Comment by pritha-srivastava** [comment]
> @pritha-srivastava +1 on adding namespace to output

@mattbenjamin : I didnt separate out the tenant and the namespace from the owner, so that the end user can use it as is in other commands.

**Comment by pritha-srivastava** [comment]
> not sure it's complete, but I favor this approach--adam, casey?

I am thinking of grouping the users under their respective clients(applications) also - such that the namespace in this case will be oidc<clientid> - so that we can have users with same uid/name belonging to different clients. Currently trying to find which is the best token claim to use as client id, since different claims represent client id in openid connect and oauth 2.0.

**Comment by mattbenjamin** [comment]
as discussed, seems like users might be used across applications, does namespacing users by client-id permit that?

**Comment by pritha-srivastava** [comment]
> as discussed, seems like users might be used across applications, does namespacing users by client-id permit that?

Hi Matt, as discussed in our one on one, uid will be unique to a user such that it can be used across applications, I didn't namespace them by client-id.

**Comment by mattbenjamin** [comment]
@adamemerson could you prioritize a review of this PR?  thanks!

**Comment by pritha-srivastava** [comment]
> It looks like this is cogent for the oidc namespace case. What is the plan for accounts that should live in other namespaces (e.g., LDAP, Swift, &c)?

I have planned to do that as part of a separate PR.

**Comment by mattbenjamin** [comment]
(this should be able to merge once we have a successful teuthology suite run)

**Comment by pritha-srivastava** [comment]
Teuthology run scheduled here: https://pulpito.ceph.com/prsrivas-2020-09-08_04:35:56-rgw-wip-rgw-sts-user-as-owner-distro-basic-smithi/

**Comment by pritha-srivastava** [comment]
> Teuthology run scheduled here: https://pulpito.ceph.com/prsrivas-2020-09-08_04:35:56-rgw-wip-rgw-sts-user-as-owner-distro-basic-smithi/

Finally got a clean run here: https://pulpito.ceph.com/prsrivas-2020-09-14_17:19:44-rgw-wip-rgw-sts-user-as-owner-distro-basic-smithi/

File: src/rgw/rgw_basic_types.h [hunk 3]
```diff
@@ -68,9 +82,18 @@ struct rgw_user {
     size_t pos = str.find('$');
     if (pos != std::string::npos) {
       tenant = str.substr(0, pos);
-      id = str.substr(pos + 1);
+      string ns_id = str.substr(pos + 1);
```
[the remark this answers is not in the document]
Comment (pritha-srivastava) [reply src/rgw/rgw_basic_types.h]: @adamemerson , @theanalyst : Please check if the changes related to string_view are correct.

**Comment by pritha-srivastava** [comment]
> > Teuthology run scheduled here: https://pulpito.ceph.com/prsrivas-2020-09-08_04:35:56-rgw-wip-rgw-sts-user-as-owner-distro-basic-smithi/
> 
> Finally got a clean run here: https://pulpito.ceph.com/prsrivas-2020-09-14_17:19:44-rgw-wip-rgw-sts-user-as-owner-distro-basic-smithi/

@mattbenjamin , @adamemerson : There are no teuthology failures and is ready to be merged.

**Comment by pritha-srivastava** [comment]
jenkins retest this please

**Comment by pritha-srivastava** [comment]
jenkins retest this please

**Comment by pritha-srivastava** [comment]
make check failed because of a failure in 'run-rbd-unit-tests-N.sh' test.

**Comment by pritha-srivastava** [comment]
jenkins test make check

**Comment by pritha-srivastava** [comment]
@mattbenjamin : All tests have passed and is ready to be merged

File: src/rgw/rgw_basic_types.h [hunk 4]
```diff
@@ -68,9 +82,19 @@ struct rgw_user {
     size_t pos = str.find('$');
     if (pos != std::string::npos) {
       tenant = str.substr(0, pos);
-      id = str.substr(pos + 1);
+      string_view sv = str;
+      string_view ns_id = sv.substr(pos + 1);
+      size_t ns_pos = ns_id.find('$');
```
Comment (cbodley) [inline src/rgw/rgw_basic_types.h:87]: won't reuse of `$` as a delimiter for namespace cause parsing ambiguities? won't `user$oidc` get parsed as tenant=user uid=oidc instead?

Comment (mattbenjamin) [reply src/rgw/rgw_basic_types.h:87]: please suggest an alternative so pritha can send another change

Comment (mattbenjamin) [reply src/rgw/rgw_basic_types.h:87]: @cbodley

Comment (mattbenjamin) [reply src/rgw/rgw_basic_types.h:87]: I didn't notice the delimiter at all, and maybe @adamemerson didn't either, so making this more salient in the code--e.g., through use of a constant--would be helpful; did we also miss parsing unit tests?

Comment (pritha-srivastava) [reply src/rgw/rgw_basic_types.h:87]: @mattbenjamin @cbodley @adamemerson : The users will be saved as tenant$oidc$uid and while parsing this all three fields will be properly parsed out. If the tenant is "" or the global tenant then the user will be saved as $oidc$uid, so in this case also all three fields will be properly parsed. For other users which are of the format tenant$uid, the logic will parse out the tenant and uid properly. Is this ok, or is there still a case that has not been taken into account?

Comment (cbodley) [reply src/rgw/rgw_basic_types.h:87]: > If the tenant is "" or the global tenant then the user will be saved as $oidc$uid, so in this case also all three fields will be properly parsed

okay, that's clever! that should be fine then

Comment (mattbenjamin) [reply src/rgw/rgw_basic_types.h:87]: that's good to hear;  what I'd add is, I do think we want to make this (parsing, parsing rules, &c) clearer in the code and maybe doc

**Comment by krasaev** [comment]
@pritha-srivastava hi there, could you tell please in what release will come these changes?

**Comment by pritha-srivastava** [comment]
> @pritha-srivastava hi there, could you tell please in what release will come these changes?

@krasaev : It should be there in the next ceph release - Pacific. There are still some changes pending related to this. https://github.com/ceph/ceph/pull/37595 is under review at the moment.

**Comment by krasaev** [comment]
@pritha-srivastava Thank you for your quick reply, have a nice day

**Pull request merged**